storage/engine/mvcc.go

```diff
@@ -860,7 +860,14 @@ func MVCCConditionalPut(engine Engine, ms *MVCCStats, key proto.Key, timestamp p
 		return err
 	}
 
-	if expValue == nil && existVal != nil {
+	if _, _, err := MVCCGet(engine, key, timestamp, true /* consistent */, nil); txn != nil && existVal != nil && bytes.Equal(value.Bytes, existVal.Bytes) &&
```

Comment:
Like most other things here, it's provisional code. I need to do a difficult dance where `TxnCoordSender` is both able to push whole batches down (it isn't yet) and single requests (which then get wrapped to single-element batches, which is the part which passes so far).
Don't get into the details of all of these things, they're going to get ripped out. Especially this part won't be necessary with your changes in.